osu.Server.Spectator/Hubs/EntityStore.cs

```diff
@@ -0,0 +1,203 @@
+    public class EntityStore<T>
+        where T : class
+    {
+        private readonly Dictionary<long, TrackedEntity> entityMapping = new Dictionary<long, TrackedEntity>();
+
+        private const int lock_timeout = 5000;
+
+        /// <summary>
+        /// Retrieve an entity with a lock for use.
+        /// </summary>
+        /// <param name="id">The ID of the requested entity.</param>
+        /// <param name="createOnMissing">Whether to create a new tracking instance if the entity is not already tracked.</param>
+        /// <returns>An <see cref="ItemUsage{T}"/> which allows reading or writing the item. This should be disposed after usage.</returns>
+        /// <exception cref="KeyNotFoundException">Thrown if <see cref="createOnMissing"/> was false and the item is not in a tracked state.</exception>
+        public async Task<ItemUsage<T>> GetForUse(long id, bool createOnMissing = false)
+        {
+            int retryCount = 10;
+
+            while (retryCount-- > 0)
+            {
+                TrackedEntity? item;
+
+                lock (entityMapping)
+                {
+                    if (!entityMapping.TryGetValue(id, out item))
+                    {
+                        if (!createOnMissing)
+                            throw new KeyNotFoundException($"Attempted to get untracked entity {typeof(T)} id {id}");
+
+                        entityMapping[id] = item = new TrackedEntity(id, this);
+                    }
+                }
+
+                try
+                {
+                    await item.ObtainLockAsync();
+                }
+                // this may be thrown if the item was destroyed between when we retrieved the item usage and took the lock.
+                catch (InvalidOperationException)
+                {
+                    // if we're looking to create on missing, we should retry the whole process now that we are aware a previous tracked instance was destroyed.
+                    if (createOnMissing)
+                        continue;
+
+                    // if we're just looking to retrieve and instance, to an external consumer, this should just be handled as the item not being tracked.
+                    throw new KeyNotFoundException($"Attempted to get untracked entity {typeof(T)} id {id}");
+                }
+
+                return new ItemUsage<T>(item);
+            }
+
+            throw new TimeoutException("Could not allocate new entity after multiple retries. Something very bad has happened");
+        }
+
+        public async Task Destroy(long id)
+        {
+            TrackedEntity? item;
+
+            lock (entityMapping)
+            {
+                if (!entityMapping.TryGetValue(id, out item))
+                    // was not tracking.
+                    return;
+            }
+
+            await item.ObtainLockAsync();
+
+            // handles removal and disposal of the semaphore.
+            item.Destroy();
```

Comment:
Maybe this should be catching (and ignoring) `InvalidOperationException` too? Hitting it would require multiple threads trying to destroy a single entity, but handling this case is simpler because receiving that exception just means someone's already done this work and there is nothing more to be done by this method.

Reply:
It should be impossible to destroy from outside a locked context (see `Destroy`'s implementation). Are you saying we should add just for added safety?

Reply:
That's not entirely what I meant - because the entity collection and entity locking are separate, there can be subtle order races between them. The scenario I'm thinking of is as follows:

```
FIRST												  SECOND

lock (entityMapping)
{
	if (!entityMapping.TryGetValue(id, out item))
		// succeeds
}

									                  lock (entityMapping)
									                  {
										                  if (!entityMapping.TryGetValue(id, out item))
											                  // succeeds
									                  }

await item.ObtainLockAsync(); // succeeds
item.Destroy();

													  await item.ObtainLockAsync();
													  // fails with InvalidOperationException
```

Reply:
Ah, you mean on the `ObtainLockAsync` line, right? Seemed like your comment was attached to `.Destroy` which confused me. Sounds valid.

Reply:
Yeah I suppose just catching on the lock obtain should suffice.